File: VocaDbModel/DataContracts/Tags/TagForEditContract.cs
#nullable disable

using System.Linq;
using System.Runtime.Serialization;
using VocaDb.Model.DataContracts.Globalization;
using VocaDb.Model.Domain.Globalization;
using VocaDb.Model.Domain.Security;
using VocaDb.Model.Domain.Tags;

namespace VocaDb.Model.DataContracts.Tags
{
	[DataContract]
	public class TagForEditContract : TagContract
	{
		public TagForEditContract()
		{
			UpdateNotes = string.Empty;
		}

		public TagForEditContract(Tag tag, bool isEmpty, IUserPermissionContext userContext)
			: base(tag, userContext.LanguagePreference)
		{
			CanDelete = EntryPermissionManager.CanDelete(userContext, tag);
			DefaultNameLanguage = tag.TranslatedName.DefaultLanguage;
			Description = new EnglishTranslatedStringContract(tag.Description);
			IsEmpty = isEmpty;
			Names = tag.Names.Select(n => new LocalizedStringWithIdContract(n)).ToArray();
			RelatedTags = tag.RelatedTags.Select(t => new TagBaseContract(t.LinkedTag, userContext.LanguagePreference, false)).ToArray();
			Thumb = (tag.Thumb != null ? new EntryThumbContract(tag.Thumb) : null);
			UpdateNotes = string.Empty;
			WebLinks = tag.WebLinks.Links.Select(w => new WebLinkContract(w)).OrderBy(w => w.DescriptionOrUrl).ToArray();
		}

		[DataMember]
		public bool CanDelete { get; init; }

		[DataMember]
		public ContentLanguageSelection DefaultNameLanguage { get; init; }

		[DataMember]
		public new EnglishTranslatedStringContract Description { get; set; }

		[DataMember]
		public bool IsEmpty { get; init; }

		[DataMember]
		public LocalizedStringWithIdContract[] Names { get; set; }

		[DataMember]
		public TagBaseContract[] RelatedTags { get; init; }

		[DataMember]
		public EntryThumbContract Thumb { get; init; }

		[DataMember]
		public string UpdateNotes { get; init; }

		[DataMember]
		public WebLinkContract[] WebLinks { get; init; }
	}
}

File: VocaDbWeb/Scripts/DataContracts/Tag/TagForEditContract.ts
import { EntryThumbContract } from '@/DataContracts/EntryThumbContract';
import { EnglishTranslatedStringContract } from '@/DataContracts/Globalization/EnglishTranslatedStringContract';
import { LocalizedStringWithIdContract } from '@/DataContracts/Globalization/LocalizedStringWithIdContract';
import { TagBaseContract } from '@/DataContracts/Tag/TagBaseContract';
import { WebLinkContract } from '@/DataContracts/WebLinkContract';
import { EntryStatus } from '@/Models/EntryStatus';
import { TagTargetTypes } from '@/Models/Tags/TagTargetTypes';

// Corresponds to the TagForEditForApiContract record class in C#.
export interface TagForEditContract {
	canDelete: boolean;
	categoryName: string;
	defaultNameLanguage: string /* TODO: enum */;
	deleted: boolean;
	description: EnglishTranslatedStringContract;
	hideFromSuggestions: boolean;
	id: number;
	mainPicture?: EntryThumbContract;
	name: string;
	names: LocalizedStringWithIdContract[];
	parent?: TagBaseContract;
	relatedTags: TagBaseContract[];
	status: EntryStatus;
	targets: TagTargetTypes;
	updateNotes: string;
	webLinks: WebLinkContract[];
}
